### PR TITLE
feat(util/find): add user/find.c and Makefile

### PR DESCRIPTION
**Pull request by LittleYe233**
Adds user/find.c and update Makefile for the "find" experiment.

### CODE REVIEW

File: user/find.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/fs.h"
+
+#define MAX_PATH_SIZE 100
+
+char *pathbase(char *path) {
```
Comment (inclyc) [inline user/find.c:8]:
```suggestion
static char *pathbase(char *path) {
```

File: user/find.c [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    static char buf[MAX_PATH_SIZE + 1];
+
+    // Find the first character after the last slash.
+    for (p = path + strlen(path); p >= path && *p != '/'; --p)
+        ;
+
+    return strcpy(buf, p + 1);
+}
+
+char *pathjoin(char *path, char *prefix, char *dst) {
```
Comment (inclyc) [inline user/find.c:19]:
```suggestion
static char *pathjoin(char *dst, const char *path, const char *prefix, ) {
```

Usually the copy destination dst is the first parameter, this is a convention. Declare untouched arrays "const" to let compilers do more assumptions.

File: user/find.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        return 1;
+    }
+
+    switch (st.type) {
+    case T_FILE:
+        if (strcmp(pathbase(path), fname) == 0) {
+            printf("%s\n", path);
+        }
+        break;
+
```
Comment (inclyc) [inline user/find.c:47]: drop \n

File: user/find.c [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    case T_DIR:
+        if (strlen(path) + 1 + DIRSIZ + 1 > MAX_PATH_SIZE) {
+            fprintf(2, "find: path %s is too long\n", path);
+            break;
+        }
+
```
Comment (inclyc) [inline user/find.c:53]: drop \n

File: user/find.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        struct dirent de;
+        while (read(fd, &de, sizeof(de)) == sizeof(de)) {
+            if (de.inum == 0 || strcmp(de.name, ".") == 0 ||
+                strcmp(de.name, "..") == 0) {
+                continue;
+            }
+
```
Comment (inclyc) [inline user/find.c:60]: drop \n

**Review by inclyc — COMMENTED** [review]
This code is very clear and easy to understand and it is correct, works fine on my machine. I left some comments about code style and conventions. This patch basically LGTM.

File: user/find.c (hunk 2 above)
Comment (inclyc) [reply user/find.c:19]: Note that `const char*` means we do not store bytes to the pointer points to, the pointer it self may change. `char * const` means the pointer address should not change, always points to the same object, but its contents may change. You should disambiguate these types:

```
const char *
char * const
const char * const
```

**Review by inclyc — APPROVED** [review]
LG

File: user/find.c (hunk 1 above)
Comment (LittleYe233) [reply user/find.c]: I notice that here `pathbase` and `pathjoin` are suggested static, but other functions in the previous experiments are not required to do so. I also searched and found something like "making a function static hides it from other translation units, which helps provide [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming))" (ref: https://stackoverflow.com/a/5319370/12002560). I was wondering that why only the two functions are asked to be static here? Or other functions don't need to be hidden from other translation units?

**Pull request merged**